madmom/features/beats.py

```diff
@@ -1238,6 +1238,380 @@ def add_arguments(parser, min_bpm=MIN_BPM, max_bpm=MAX_BPM,
+            if activation > self.THRESHOLD:
+                self.beats = self.beats + [idx]
+
+        def process(self, activations, idx):
+            """
+            Set the next prediction, score the agent and create
+            new child agents if necessary. This method is called after
+            the outer window has passed for offline and online.
+
+            Parameters
+            ----------
+            activations : list
+                Activation window which surrounds the prediction by
+                outer window length on both sides. For online mode we wait
+                until all of that information is available before calling
+                this method.
+            idx : int
+                Global frame counter index of the last activation. Since we
+                calculate everything in absolute frame times.
+
+            Returns
+            -------
+            agents : list
+                New child agent objects
+
+            """
+            # calculate frames to look around
+            frames = int(self.interval * self.OUTER_WINDOW)
+            # get predicted activation
+            act = activations[int(idx - self.prediction)]
+            # get max index within outer window
+            max_idx = idx - len(activations) + np.argmax(activations)
+            # distance between max activation and predicted position
+            error = max_idx - self.prediction
+            # faster agents should not get a better score
+            normalization = self.interval / self._MIN_INTERVAL
+            # if max activation was in inner window
+            if abs(error) <= self.INNER_WINDOW:
+                # if no beat has been accepted yet, accept max (for offline)
+                if self.beats[-1] < idx - len(activations):
+                    self.accept(max(activations), max_idx)
+                # update prediction
+                self.prediction = max_idx + self.interval
+                # reward agent for detecting the beat
+                self.score += (1 - abs(error) / frames) * normalization * act
+                # adapt agent to error
+                self.interval += int(error * self.CORRECTION_FACTOR)
+                self.prediction += int(error * self.CORRECTION_FACTOR)
+                # return no new child agents
+                return []
+            # if max activation was in outer window
+            else:
+                # if no beat has been accepted yet, accept act (for offline)
+                if self.beats[-1] < idx - len(activations):
+                    self.accept(act, self.prediction)
+                # update prediction
+                self.prediction += self.interval
+                # create child agents
+                new_agents = self.fork(error)
+                # penalize agent for not detecting the beat
+                self.score -= (abs(error) / frames) * normalization * act
+                # return new child agents
+                return new_agents
```

Comment:
This could probably be refactored a bit by moving common code outside the if/else clause and then decide later on whether new agents need to be created or not.